sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,426 @@
+                // TODO: signal and cleanup
+                panic!("ERROR: Received HELLO when not registering");
+            }
+            self.update_state(AppState::ServerRegistered);
+            setup_call(&self.app_state, &self.out);
+            return Ok(());
+        }
+        if msg_text == "SESSION_OK" {
+            if self.app_state.load(Ordering::Relaxed) != (AppState::PeerConnecting as usize) {
+                // TODO: signal and cleanup
+                panic!("ERROR: Received SESSION_OK when not calling");
+            }
+            self.update_state(AppState::PeerConnected);
+            self.webrtc = Some(start_pipeline(&self.app_state, &self.out));
+            return Ok(());
+        }
+
+        if msg_text.starts_with("ERROR") {
+            println!("Got error message! {}", msg_text);
+            let error = match self.app_state.load(Ordering::Relaxed) {
+                x if x == AppState::ServerConnecting as usize => AppState::ServerConnectionError,
+                x if x == AppState::ServerRegistering as usize => AppState::ServerRegisteringError,
+                x if x == AppState::PeerConnecting as usize => AppState::PeerConnectionError,
+                x if x == AppState::PeerConnected as usize => AppState::PeerCallError,
+                x if x == AppState::PeerCallNegotiating as usize => AppState::PeerCallError,
+                _ => AppState::AppStateErr,
+            };
+            panic!("Got websocket error {:?}", error);
+            // TODO: signal & cleanup
+        }
+
+        let json_msg: serde_json::Value = serde_json::from_str(&msg_text).unwrap();
+        if json_msg.get("sdp").is_some() {
+            assert_eq!(
+                self.app_state.load(Ordering::Relaxed),
+                (AppState::PeerCallNegotiating as usize)
+            );
+
+            if !json_msg["sdp"].get("type").is_some() {
+                println!("ERROR: received SDP without 'type'");
+                return Ok(());
+            }
+            let sdptype = &json_msg["sdp"]["type"];
+            assert_eq!(sdptype, "answer");
+            let text = &json_msg["sdp"]["sdp"];
+            print!("Received answer:\n{}\n", text.as_str().unwrap());
+
+            let ret =
+                gstreamer_sdp::SDPMessage::parse_buffer(text.as_str().unwrap().as_bytes()).unwrap();
+            let answer = gstreamer_webrtc::WebRTCSessionDescription::new(
+                gstreamer_webrtc::WebRTCSDPType::Answer,
+                ret,
+            );
+            let promise = gst::Promise::new();
+            self.webrtc
+                .as_ref()
+                .unwrap()
+                .emit(
+                    "set-remote-description",
+                    &[&answer.to_value(), &promise.to_value()],
+                )
+                .unwrap();
+            self.update_state(AppState::PeerCallStarted);
+        }
+        if json_msg.get("ice").is_some() {
```

Comment:
Maybe we could use `serde-json` here for directly converting between a Rust struct and JSON?